# Toronto neighborhoods

**In this assignment, you will be required to explore, segment, and cluster the neighborhoods in the city of Toronto.** 

The neighborhood data for Toronto is not readily available on the internet. However, a Wikipedia page exists that has all the information we need to explore and cluster the neighborhoods in Toronto. 

According to above, this assignment will consist of three notebooks, each of one for a different part: 

- **Parts 1 and 2**: you will be required to get the data, clean it, and then read it into a pandas dataframe so that it is in a suitable structured format. 

- **Part 3**: once the data is in a structured format, you can analyze the dataset to explore and cluster the neighborhoods in the city of Toronto.

Let's start!

In [1]:
#!conda install -c conda-forge geopy --yes           # uncomment only if it is necessary to install the 'geopy' library

Import the necessary libraries to run this notebook:

In [2]:
                     import numpy             as np      # library to handle data in a vectorized manner
                     import pandas            as pd      # library for data analysis
from pandas.io.json  import json_normalize               # tranform JSON file into a pandas dataframe    
from geopy.geocoders import Nominatim                    # convert an address into latitude and longitude values
                     import folium                       # map rendering library
                     import requests                     # library to handle requests
from sklearn.cluster import KMeans                       # import k-means from clustering stage
                     import matplotlib.cm     as cm      # matplotlib and associated plotting modules
                     import matplotlib.colors as colors  # matplotlib and associated plotting modules

## Part 1: Postal codes

**Scrape the Wikipedia page https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M in order**

**1) to obtain the data that is in the table of postal codes and** 

**2) to transform the data into a pandas dataframe of the postal code of each neighborhood along with the borough name and neighborhood name.**

## 1.1. Obtain the data

**Use the BeautifulSoup package or *any other way* you are comfortable with to transform the data in the table on the Wikipedia page into the pandas dataframe.**

Instead of BeautifulSoup package, I preferred to use the function *'read_html'* from Pandas to read the HTML table into a dataframe.

In [3]:
df_cp = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')[0]

Explore the data:

In [4]:
df_cp.shape         # dataframe dimension (rows, columns)

(288, 3)

In [5]:
df_cp.columns       # column names

Index(['Postcode', 'Borough', 'Neighbourhood'], dtype='object')

In [6]:
df_cp               # dataframe display

Postcode           Borough  \
0        M1A      Not assigned   
1        M2A      Not assigned   
2        M3A        North York   
3        M4A        North York   
4        M5A  Downtown Toronto   
5        M5A  Downtown Toronto   
6        M6A        North York   
7        M6A        North York   
8        M7A      Queen's Park   
9        M8A      Not assigned   
10       M9A         Etobicoke   
11       M1B       Scarborough   
12       M1B       Scarborough   
13       M2B      Not assigned   
14       M3B        North York   
15       M4B         East York   
16       M4B         East York   
17       M5B  Downtown Toronto   
18       M5B  Downtown Toronto   
19       M6B        North York   
20       M7B      Not assigned   
21       M8B      Not assigned   
22       M9B         Etobicoke   
23       M9B         Etobicoke   
24       M9B         Etobicoke   
25       M9B         Etobicoke   
26       M9B         Etobicoke   
27       M1C       Scarborough   
28       M1C       Scarborough   
29       M1C       Scarborough   
..       ...               ...   
258      M9X      Not assigned   
259      M1Y      Not assigned   
260      M2Y      Not assigned   
261      M3Y      Not assigned   
262      M4Y  Downtown Toronto   
263      M5Y      Not assigned   
264      M6Y      Not assigned   
265      M7Y      East Toronto   
266      M8Y         Etobicoke   
267      M8Y         Etobicoke   
268      M8Y         Etobicoke   
269      M8Y         Etobicoke   
270      M8Y         Etobicoke   
271      M8Y         Etobicoke   
272      M8Y         Etobicoke   
273      M8Y         Etobicoke   
274      M9Y      Not assigned   
275      M1Z      Not assigned   
276      M2Z      Not assigned   
277      M3Z      Not assigned   
278      M4Z      Not assigned   
279      M5Z      Not assigned   
280      M6Z      Not assigned   
281      M7Z      Not assigned   
282      M8Z         Etobicoke   
283      M8Z         Etobicoke   
284      M8Z         Etobicoke   
285      M8Z         Etobicoke   
286      M8Z         Etobicoke   
287      M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                                         Harbourfront  
5                                          Regent Park  
6                                     Lawrence Heights  
7                                       Lawrence Manor  
8                                         Not assigned  
9                                         Not assigned  
10                                    Islington Avenue  
11                                               Rouge  
12                                             Malvern  
13                                        Not assigned  
14                                     Don Mills North  
15                                    Woodbine Gardens  
16                                       Parkview Hill  
17                                             Ryerson  
18                                     Garden District  
19                                           Glencairn  
20                                        Not assigned  
21                                        Not assigned  
22                                          Cloverdale  
23                                           Islington  
24                                        Martin Grove  
25                                    Princess Gardens  
26                                     West Deane Park  
27                                      Highland Creek  
28                                          Rouge Hill  
29                                          Port Union  
..                                                 ...  
258                                       Not assigned  
259           

## 1.2. Transform the data

**1.2.1.The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood.**

Rename the column names to match the ones in the required dataframe:

In [7]:
df_cp.rename(columns={'Postcode':'PostalCode', 'Neighbourhood':'Neighborhood'}, inplace=True)   # columns rename 
df_cp.columns                                                                                   # verification

Index(['PostalCode', 'Borough', 'Neighborhood'], dtype='object')

**1.2.2. Only process the cells that have an assigned borough. Ignore cells with a borough that is *'Not assigned'*.**

In [8]:
filter_1 = (df_cp['Borough']!='Not assigned') # create a filter with the condition 'Borough' different to'Not assigned'
df_cp = df_cp[filter_1]                       # apply the filter 
'Not assigned' is df_cp['Borough']            # verification

False

In [9]:
df_cp                                         # visualize the resulting dataframe

PostalCode           Borough  \
2          M3A        North York   
3          M4A        North York   
4          M5A  Downtown Toronto   
5          M5A  Downtown Toronto   
6          M6A        North York   
7          M6A        North York   
8          M7A      Queen's Park   
10         M9A         Etobicoke   
11         M1B       Scarborough   
12         M1B       Scarborough   
14         M3B        North York   
15         M4B         East York   
16         M4B         East York   
17         M5B  Downtown Toronto   
18         M5B  Downtown Toronto   
19         M6B        North York   
22         M9B         Etobicoke   
23         M9B         Etobicoke   
24         M9B         Etobicoke   
25         M9B         Etobicoke   
26         M9B         Etobicoke   
27         M1C       Scarborough   
28         M1C       Scarborough   
29         M1C       Scarborough   
31         M3C        North York   
32         M3C        North York   
33         M4C         East York   
34         M5C  Downtown Toronto   
35         M6C              York   
38         M9C         Etobicoke   
..         ...               ...   
235        M9V         Etobicoke   
236        M1W       Scarborough   
239        M4W  Downtown Toronto   
240        M5W  Downtown Toronto   
243        M8W         Etobicoke   
244        M8W         Etobicoke   
245        M9W         Etobicoke   
246        M1X       Scarborough   
249        M4X  Downtown Toronto   
250        M4X  Downtown Toronto   
251        M5X  Downtown Toronto   
252        M5X  Downtown Toronto   
255        M8X         Etobicoke   
256        M8X         Etobicoke   
257        M8X         Etobicoke   
262        M4Y  Downtown Toronto   
265        M7Y      East Toronto   
266        M8Y         Etobicoke   
267        M8Y         Etobicoke   
268        M8Y         Etobicoke   
269        M8Y         Etobicoke   
270        M8Y         Etobicoke   
271        M8Y         Etobicoke   
272        M8Y         Etobicoke   
273        M8Y         Etobicoke   
282        M8Z         Etobicoke   
283        M8Z         Etobicoke   
284        M8Z         Etobicoke   
285        M8Z         Etobicoke   
286        M8Z         Etobicoke   

                                          Neighborhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                                         Harbourfront  
5                                          Regent Park  
6                                     Lawrence Heights  
7                                       Lawrence Manor  
8                                         Not assigned  
10                                    Islington Avenue  
11                                               Rouge  
12                                             Malvern  
14                                     Don Mills North  
15                                    Woodbine Gardens  
16                                       Parkview Hill  
17                                             Ryerson  
18                                     Garden District  
19                                           Glencairn  
22                                          Cloverdale  
23                                           Islington  
24                                        Martin Grove  
25                                    Princess Gardens  
26                                     West Deane Park  
27                                      Highland Creek  
28                                          Rouge Hill  
29                                          Port Union  
31                                     Flemingdon Park  
32                                     Don Mills South  
33                                    Woodbine Heights  
34                                      St. James Town  
35                                  Humewood-Cedarvale  
38                                   Bloordale Gardens  
..  

**1.2.3. If a cell has a borough but a *'Not assigned'* neighborhood, then the neighborhood will be the same as the borough.** 

In [10]:
filter_21 = (df_cp['Neighborhood'] == 'Not assigned')   # create a filter
labels_21 = df_cp[filter_21].index                       # apply the filter and get the row labels
df_cp[filter_21]                                        # visualize the filtered dataframe

PostalCode       Borough  Neighborhood
8        M7A  Queen's Park  Not assigned

There is only one row where the value of **'Neighborhood'** is *'Not assigned'*. For this row, we equal the Borough and Neighborhood columns.

In [11]:
# Change 'Neighborhood' if it is 'Not assigned'
# 1 - Valid only in this case, where filter_2 only has 1 element.
#df_cp.loc[rows_2[0]]['Neighbourhood'] = df_cp.loc[rows_2[0]]['Borough']  # df.loc - access to a row by label

# 2 - Valid in general, for any filter size.
for i in labels_21:                                             
    df_cp.loc[i]['Neighborhood'] = df_cp.loc[i]['Borough'] 

Validation (three different ways):

In [12]:
print('Not assigned' in df_cp['Neighborhood'])           # first

False


In [13]:
filter_22 = (df_cp['Neighborhood'] == 'Not assigned')    # second      
df_cp[filter_22]

Empty DataFrame
Columns: [PostalCode, Borough, Neighborhood]
Index: []

In [14]:
df_cp.loc[[8]]                                            # third

PostalCode       Borough  Neighborhood
8        M7A  Queen's Park  Queen's Park

The value of **'Neigborhood'** is the same than the value of **'Borough'** in the row with label 8. That's right!

**1.2.4. When more than one neighborhood exists in one postal code area, these rows will be combined into one row with the neighborhoods separated with a comma.**

In [15]:
df_cp             # display the dataframe before combination

PostalCode           Borough  \
2          M3A        North York   
3          M4A        North York   
4          M5A  Downtown Toronto   
5          M5A  Downtown Toronto   
6          M6A        North York   
7          M6A        North York   
8          M7A      Queen's Park   
10         M9A         Etobicoke   
11         M1B       Scarborough   
12         M1B       Scarborough   
14         M3B        North York   
15         M4B         East York   
16         M4B         East York   
17         M5B  Downtown Toronto   
18         M5B  Downtown Toronto   
19         M6B        North York   
22         M9B         Etobicoke   
23         M9B         Etobicoke   
24         M9B         Etobicoke   
25         M9B         Etobicoke   
26         M9B         Etobicoke   
27         M1C       Scarborough   
28         M1C       Scarborough   
29         M1C       Scarborough   
31         M3C        North York   
32         M3C        North York   
33         M4C         East York   
34         M5C  Downtown Toronto   
35         M6C              York   
38         M9C         Etobicoke   
..         ...               ...   
235        M9V         Etobicoke   
236        M1W       Scarborough   
239        M4W  Downtown Toronto   
240        M5W  Downtown Toronto   
243        M8W         Etobicoke   
244        M8W         Etobicoke   
245        M9W         Etobicoke   
246        M1X       Scarborough   
249        M4X  Downtown Toronto   
250        M4X  Downtown Toronto   
251        M5X  Downtown Toronto   
252        M5X  Downtown Toronto   
255        M8X         Etobicoke   
256        M8X         Etobicoke   
257        M8X         Etobicoke   
262        M4Y  Downtown Toronto   
265        M7Y      East Toronto   
266        M8Y         Etobicoke   
267        M8Y         Etobicoke   
268        M8Y         Etobicoke   
269        M8Y         Etobicoke   
270        M8Y         Etobicoke   
271        M8Y         Etobicoke   
272        M8Y         Etobicoke   
273        M8Y         Etobicoke   
282        M8Z         Etobicoke   
283        M8Z         Etobicoke   
284        M8Z         Etobicoke   
285        M8Z         Etobicoke   
286        M8Z         Etobicoke   

                                          Neighborhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                                         Harbourfront  
5                                          Regent Park  
6                                     Lawrence Heights  
7                                       Lawrence Manor  
8                                         Queen's Park  
10                                    Islington Avenue  
11                                               Rouge  
12                                             Malvern  
14                                     Don Mills North  
15                                    Woodbine Gardens  
16                                       Parkview Hill  
17                                             Ryerson  
18                                     Garden District  
19                                           Glencairn  
22                                          Cloverdale  
23                                           Islington  
24                                        Martin Grove  
25                                    Princess Gardens  
26                                     West Deane Park  
27                                      Highland Creek  
28                                          Rouge Hill  
29                                          Port Union  
31                                     Flemingdon Park  
32                                     Don Mills South  
33                                    Woodbine Heights  
34                                      St. James Town  
35                                  Humewood-Cedarvale  
38                                   Bloordale Gardens  
..  

In [16]:
print('The city of Toronto has {} boroughs, {} postal codes and {} neighborhoods.'.format(
       len(df_cp['Borough'].unique()),len(df_cp['PostalCode'].unique()), df_cp.shape[0]))

The city of Toronto has 11 boroughs, 103 postal codes and 211 neighborhoods.


Get the row labels for the postal codes with more than one neighborhood:

In [17]:
filter_3 = df_cp[['PostalCode']].duplicated()
labels_3 = df_cp[filter_3].index
labels_3

Int64Index([  5,   7,  12,  16,  18,  23,  24,  25,  26,  28,
            ...
            268, 269, 270, 271, 272, 273, 283, 284, 285, 286],
           dtype='int64', length=108)

Combine the neighborhoods for these postal codes into one row:

In [18]:
for k in range(labels_3.size):
    neigh = df_cp.loc[labels_3[k]]['Neighborhood']  # get neighborhood to combine
    row = df_cp.index.get_loc(labels_3[k])          # get the row label origen
    df_cp.iloc[row-1]['Neighborhood'] = df_cp.iloc[row-1]['Neighborhood'] + ',' + neigh   # combine
    df_cp = df_cp.drop(labels_3[k], axis=0)         # delete original row

In [19]:
df_cp          # validation

PostalCode           Borough  \
2          M3A        North York   
3          M4A        North York   
4          M5A  Downtown Toronto   
6          M6A        North York   
8          M7A      Queen's Park   
10         M9A         Etobicoke   
11         M1B       Scarborough   
14         M3B        North York   
15         M4B         East York   
17         M5B  Downtown Toronto   
19         M6B        North York   
22         M9B         Etobicoke   
27         M1C       Scarborough   
31         M3C        North York   
33         M4C         East York   
34         M5C  Downtown Toronto   
35         M6C              York   
38         M9C         Etobicoke   
42         M1E       Scarborough   
47         M4E      East Toronto   
48         M5E  Downtown Toronto   
49         M6E              York   
53         M1G       Scarborough   
56         M4G         East York   
57         M5G  Downtown Toronto   
58         M6G  Downtown Toronto   
62         M1H       Scarborough   
63         M2H        North York   
64         M3H        North York   
67         M4H         East York   
..         ...               ...   
168        M4R   Central Toronto   
169        M5R   Central Toronto   
172        M6R      West Toronto   
174        M7R       Mississauga   
176        M9R         Etobicoke   
180        M1S       Scarborough   
183        M4S   Central Toronto   
184        M5S  Downtown Toronto   
186        M6S      West Toronto   
191        M1T       Scarborough   
196        M4T   Central Toronto   
198        M5T  Downtown Toronto   
205        M1V       Scarborough   
211        M4V   Central Toronto   
216        M5V  Downtown Toronto   
225        M8V         Etobicoke   
228        M9V         Etobicoke   
236        M1W       Scarborough   
239        M4W  Downtown Toronto   
240        M5W  Downtown Toronto   
243        M8W         Etobicoke   
245        M9W         Etobicoke   
246        M1X       Scarborough   
249        M4X  Downtown Toronto   
251        M5X  Downtown Toronto   
255        M8X         Etobicoke   
262        M4Y  Downtown Toronto   
265        M7Y      East Toronto   
266        M8Y         Etobicoke   
282        M8Z         Etobicoke   

                                          Neighborhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                             Harbourfront,Regent Park  
6                      Lawrence Heights,Lawrence Manor  
8                                         Queen's Park  
10                                    Islington Avenue  
11                                       Rouge,Malvern  
14                                     Don Mills North  
15                      Woodbine Gardens,Parkview Hill  
17                             Ryerson,Garden District  
19                                           Glencairn  
22   Cloverdale,Islington,Martin Grove,Princess Gar...  
27                Highland Creek,Rouge Hill,Port Union  
31                     Flemingdon Park,Don Mills South  
33                                    Woodbine Heights  
34                                      St. James Town  
35                                  Humewood-Cedarvale  
38   Bloordale Gardens,Eringate,Markland Wood,Old B...  
42                     Guildwood,Morningside,West Hill  
47                                         The Beaches  
48                                         Berczy Park  
49                                 Caledonia-Fairbanks  
53                                              Woburn  
56                                             Leaside  
57                                  Central Bay Street  
58                                            Christie  
62                                           Cedarbrae  
63                                   Hillcrest Village  
64       Bathurst Manor,Downsview North,Wilson Heights  
67                                    Thorncliffe Park  
..  

**1.2.5. Change the row labels to match the ones in the required dataframe.**

In [20]:
df_cp = df_cp.reset_index(drop = True)        # rows rename 
df_cp                                         # verification

PostalCode           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A      Queen's Park   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   
12         M1C       Scarborough   
13         M3C        North York   
14         M4C         East York   
15         M5C  Downtown Toronto   
16         M6C              York   
17         M9C         Etobicoke   
18         M1E       Scarborough   
19         M4E      East Toronto   
20         M5E  Downtown Toronto   
21         M6E              York   
22         M1G       Scarborough   
23         M4G         East York   
24         M5G  Downtown Toronto   
25         M6G  Downtown Toronto   
26         M1H       Scarborough   
27         M2H        North York   
28         M3H        North York   
29         M4H         East York   
..         ...               ...   
73         M4R   Central Toronto   
74         M5R   Central Toronto   
75         M6R      West Toronto   
76         M7R       Mississauga   
77         M9R         Etobicoke   
78         M1S       Scarborough   
79         M4S   Central Toronto   
80         M5S  Downtown Toronto   
81         M6S      West Toronto   
82         M1T       Scarborough   
83         M4T   Central Toronto   
84         M5T  Downtown Toronto   
85         M1V       Scarborough   
86         M4V   Central Toronto   
87         M5V  Downtown Toronto   
88         M8V         Etobicoke   
89         M9V         Etobicoke   
90         M1W       Scarborough   
91         M4W  Downtown Toronto   
92         M5W  Downtown Toronto   
93         M8W         Etobicoke   
94         M9W         Etobicoke   
95         M1X       Scarborough   
96         M4X  Downtown Toronto   
97         M5X  Downtown Toronto   
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                             Harbourfront,Regent Park  
3                      Lawrence Heights,Lawrence Manor  
4                                         Queen's Park  
5                                     Islington Avenue  
6                                        Rouge,Malvern  
7                                      Don Mills North  
8                       Woodbine Gardens,Parkview Hill  
9                              Ryerson,Garden District  
10                                           Glencairn  
11   Cloverdale,Islington,Martin Grove,Princess Gar...  
12                Highland Creek,Rouge Hill,Port Union  
13                     Flemingdon Park,Don Mills South  
14                                    Woodbine Heights  
15                                      St. James Town  
16                                  Humewood-Cedarvale  
17   Bloordale Gardens,Eringate,Markland Wood,Old B...  
18                     Guildwood,Morningside,West Hill  
19                                         The Beaches  
20                                         Berczy Park  
21                                 Caledonia-Fairbanks  
22                                              Woburn  
23                                             Leaside  
24                                  Central Bay Street  
25                                            Christie  
26                                           Cedarbrae  
27                                   Hillcrest Village  
28       Bathurst Manor,Downsview North,Wilson Heights  
29                                    Thorncliffe Park  
..  

**1.2.6. In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe.**

In [21]:
print('The number of rows of my postal codes dataframe is: ', df_cp.shape[0])

The number of rows of my postal codes dataframe is:  103


## Part 2: Geographical coordinates of each postal code

**We need to get the latitude and the longitude coordinates of each neighborhood in order to utilize the Foursquare location data. We can use the Geocoder Python package or a csv file that has the geographical coordinates of each postal code to create the solicited dataframe.**

## 2.1. Obtain the data

I got the geographical coordinates from the available file. Here is a link to this csv file : http://cocl.us/Geospatial_data.

In [22]:
df_gc = pd.read_csv('http://cocl.us/Geospatial_data')

Explore the coordinates dataframe:

In [23]:
df_gc.shape      # dataframe dimensions (rows, columns)

(103, 3)

The geographical coordinates dataframe has the same number of rows than my postal codes dataframe. That's right!

In [24]:
df_gc.columns   # columns labels

Index(['Postal Code', 'Latitude', 'Longitude'], dtype='object')

In [25]:
df_gc.head()    # the first 5 rows

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

## 2.2. Transform the data

**2.2.1. Change the name of the first column in the coordinates dataframe to match the one in the postal codes dataframe to be able to join both dataframes.**

In [26]:
df_gc.rename(columns = {'Postal Code':'PostalCode'}, inplace = True)

In [27]:
df_gc.columns

Index(['PostalCode', 'Latitude', 'Longitude'], dtype='object')

**2.2.2. Join the postal codes and the geographical coordinates dataframes to obtain the requerided dataframe to segmentation and clustering.**

In [28]:
df_tn = df_cp.merge(df_gc, on = 'PostalCode')

The solicited datraframe to segmentation and clustering is: 

In [29]:
df_tn

PostalCode           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A      Queen's Park   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   
12         M1C       Scarborough   
13         M3C        North York   
14         M4C         East York   
15         M5C  Downtown Toronto   
16         M6C              York   
17         M9C         Etobicoke   
18         M1E       Scarborough   
19         M4E      East Toronto   
20         M5E  Downtown Toronto   
21         M6E              York   
22         M1G       Scarborough   
23         M4G         East York   
24         M5G  Downtown Toronto   
25         M6G  Downtown Toronto   
26         M1H       Scarborough   
27         M2H        North York   
28         M3H        North York   
29         M4H         East York   
..         ...               ...   
73         M4R   Central Toronto   
74         M5R   Central Toronto   
75         M6R      West Toronto   
76         M7R       Mississauga   
77         M9R         Etobicoke   
78         M1S       Scarborough   
79         M4S   Central Toronto   
80         M5S  Downtown Toronto   
81         M6S      West Toronto   
82         M1T       Scarborough   
83         M4T   Central Toronto   
84         M5T  Downtown Toronto   
85         M1V       Scarborough   
86         M4V   Central Toronto   
87         M5V  Downtown Toronto   
88         M8V         Etobicoke   
89         M9V         Etobicoke   
90         M1W       Scarborough   
91         M4W  Downtown Toronto   
92         M5W  Downtown Toronto   
93         M8W         Etobicoke   
94         M9W         Etobicoke   
95         M1X       Scarborough   
96         M4X  Downtown Toronto   
97         M5X  Downtown Toronto   
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                             Harbourfront,Regent Park  43.654260 -79.360636  
3                      Lawrence Heights,Lawrence Manor  43.718518 -79.464763  
4                                         Queen's Park  43.662301 -79.389494  
5                                     Islington Avenue  43.667856 -79.532242  
6                                        Rouge,Malvern  43.806686 -79.194353  
7                                      Don Mills North  43.745906 -79.352188  
8                       Woodbine Gardens,Parkview Hill  43.706397 -79.309937  
9                              Ryerson,Garden District  43.657162 -79.378937  
10                                           Glencairn  43.709577 -79.445073  
11   Cloverdale,Islington,Martin Grove,Princess Gar...  43.650943 -79.554724  
12                Highland Creek,Rouge Hill,Port Union  43.784535 -79.160497  
13                     Flemingdon Park,Don Mills South  43.725900 -79.340923  
14                                    Woodbine Heights  43.695344 -79.318389  
15                                      St. James Town  43.651494 -79.375418  
16                                  Humewood-Cedarvale  43.693781 -79.428191  
17   Bloordale Gardens,Eringate,Markland Wood,Old B...  43.643515 -79.577201  
18                     Guildwood,Morningside,West Hill  43.763573 -79.188711  
19                                         The Beaches  43.676357 -79.293031  
20                                         Berczy Park  43.644771 -79.373306  
21                               

## Part 3: Segmentation, clustering and visualization

**Explore and cluster the neighborhoods in Toronto using the Foursquare location data. You can decide to work with only boroughs that contain the word Toronto.**

You will get the most common venue categories in each neighborhood, and then use this feature to group the neighborhoods into clusters. You will use the *k*-means clustering algorithm to complete this task. Finally, you will use the Folium library to visualize the neighborhoods in Toronto and their emerging clusters.

## 3.1. Visualization of the neighborhoods of the city of Toronto

 Create a map of Toronto with neighborhoods superimposed on top.


**3.1.1. Get the latitude and longitude values of Toronto**

In [30]:
address = 'Toronto'

geolocator = Nominatim(user_agent = "toronto_explorer")
location   = geolocator.geocode(address)
latitude   = location.latitude
longitude  = location.longitude

print('The geographical coordinates of the city of Toronto are latitude = {} and longitude = {}.'.format(latitude, longitude))

The geographical coordinates of the city of Toronto are latitude = 43.653963 and longitude = -79.387207.


**3.1.2. Create a map of Toronto neighborhoods**

From **Part 1** (step 1.2.4), we know **the city of Toronto has 11 boroughs, 103 postal codes and 211 neighborhoods**.

In [31]:
df_tn         # the solicited datraframe for segmentation and clustering 

PostalCode           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A      Queen's Park   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   
12         M1C       Scarborough   
13         M3C        North York   
14         M4C         East York   
15         M5C  Downtown Toronto   
16         M6C              York   
17         M9C         Etobicoke   
18         M1E       Scarborough   
19         M4E      East Toronto   
20         M5E  Downtown Toronto   
21         M6E              York   
22         M1G       Scarborough   
23         M4G         East York   
24         M5G  Downtown Toronto   
25         M6G  Downtown Toronto   
26         M1H       Scarborough   
27         M2H        North York   
28         M3H        North York   
29         M4H         East York   
..         ...               ...   
73         M4R   Central Toronto   
74         M5R   Central Toronto   
75         M6R      West Toronto   
76         M7R       Mississauga   
77         M9R         Etobicoke   
78         M1S       Scarborough   
79         M4S   Central Toronto   
80         M5S  Downtown Toronto   
81         M6S      West Toronto   
82         M1T       Scarborough   
83         M4T   Central Toronto   
84         M5T  Downtown Toronto   
85         M1V       Scarborough   
86         M4V   Central Toronto   
87         M5V  Downtown Toronto   
88         M8V         Etobicoke   
89         M9V         Etobicoke   
90         M1W       Scarborough   
91         M4W  Downtown Toronto   
92         M5W  Downtown Toronto   
93         M8W         Etobicoke   
94         M9W         Etobicoke   
95         M1X       Scarborough   
96         M4X  Downtown Toronto   
97         M5X  Downtown Toronto   
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                             Harbourfront,Regent Park  43.654260 -79.360636  
3                      Lawrence Heights,Lawrence Manor  43.718518 -79.464763  
4                                         Queen's Park  43.662301 -79.389494  
5                                     Islington Avenue  43.667856 -79.532242  
6                                        Rouge,Malvern  43.806686 -79.194353  
7                                      Don Mills North  43.745906 -79.352188  
8                       Woodbine Gardens,Parkview Hill  43.706397 -79.309937  
9                              Ryerson,Garden District  43.657162 -79.378937  
10                                           Glencairn  43.709577 -79.445073  
11   Cloverdale,Islington,Martin Grove,Princess Gar...  43.650943 -79.554724  
12                Highland Creek,Rouge Hill,Port Union  43.784535 -79.160497  
13                     Flemingdon Park,Don Mills South  43.725900 -79.340923  
14                                    Woodbine Heights  43.695344 -79.318389  
15                                      St. James Town  43.651494 -79.375418  
16                                  Humewood-Cedarvale  43.693781 -79.428191  
17   Bloordale Gardens,Eringate,Markland Wood,Old B...  43.643515 -79.577201  
18                     Guildwood,Morningside,West Hill  43.763573 -79.188711  
19                                         The Beaches  43.676357 -79.293031  
20                                         Berczy Park  43.644771 -79.373306  
21                               

In [32]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_tn['Latitude'], df_tn['Longitude'], df_tn['Borough'], df_tn['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

## 3.2. Segmentation and clustering of only the neighborhoods corresponding to boroughs that contain the word 'Toronto'

I decided to work with only boroughs that contain the word Toronto. 

### 3.2.1. Obtain the data

We visualize the boroughs of Toronto with the number of postal codes associated with them.

In [33]:
df_tn.groupby('Borough').count()[['PostalCode']]

PostalCode
Borough                     
Central Toronto            9
Downtown Toronto          18
East Toronto               5
East York                  5
Etobicoke                 12
Mississauga                1
North York                24
Queen's Park               1
Scarborough               17
West Toronto               6
York                       5

We will segment and cluster only the neighborhoods corresponding to the following boroughs:
 - **Central Toronto**
 - **Downtown Toronto**
 - **East Toronto**
 - **West Toronto**

So let's slice the dataframe for the city of Toronto and create a new dataframe of these four borouhgs:

In [34]:
filter_ = (df_tn['Borough'] == ('Central Toronto')) | (df_tn['Borough'] == ('Downtown Toronto'))| (df_tn['Borough'] == ('East Toronto'))| (df_tn['Borough'] == ('West Toronto'))
df_t_cp = df_tn[filter_].reset_index(drop=True)
df_t_cp

PostalCode           Borough  \
0         M5A  Downtown Toronto   
1         M5B  Downtown Toronto   
2         M5C  Downtown Toronto   
3         M4E      East Toronto   
4         M5E  Downtown Toronto   
5         M5G  Downtown Toronto   
6         M6G  Downtown Toronto   
7         M5H  Downtown Toronto   
8         M6H      West Toronto   
9         M5J  Downtown Toronto   
10        M6J      West Toronto   
11        M4K      East Toronto   
12        M5K  Downtown Toronto   
13        M6K      West Toronto   
14        M4L      East Toronto   
15        M5L  Downtown Toronto   
16        M4M      East Toronto   
17        M4N   Central Toronto   
18        M5N   Central Toronto   
19        M4P   Central Toronto   
20        M5P   Central Toronto   
21        M6P      West Toronto   
22        M4R   Central Toronto   
23        M5R   Central Toronto   
24        M6R      West Toronto   
25        M4S   Central Toronto   
26        M5S  Downtown Toronto   
27        M6S      West Toronto   
28        M4T   Central Toronto   
29        M5T  Downtown Toronto   
30        M4V   Central Toronto   
31        M5V  Downtown Toronto   
32        M4W  Downtown Toronto   
33        M5W  Downtown Toronto   
34        M4X  Downtown Toronto   
35        M5X  Downtown Toronto   
36        M4Y  Downtown Toronto   
37        M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  
0                            Harbourfront,Regent Park  43.654260 -79.360636  
1                             Ryerson,Garden District  43.657162 -79.378937  
2                                      St. James Town  43.651494 -79.375418  
3                                         The Beaches  43.676357 -79.293031  
4                                         Berczy Park  43.644771 -79.373306  
5                                  Central Bay Street  43.657952 -79.387383  
6                                            Christie  43.669542 -79.422564  
7                              Adelaide,King,Richmond  43.650571 -79.384568  
8                         Dovercourt Village,Dufferin  43.669005 -79.442259  
9     Harbourfront East,Toronto Islands,Union Station  43.640816 -79.381752  
10                            Little Portugal,Trinity  43.647927 -79.419750  
11                        The Danforth West,Riverdale  43.679557 -79.352188  
12            Design Exchange,Toronto Dominion Centre  43.647177 -79.381576  
13         Brockton,Exhibition Place,Parkdale Village  43.636847 -79.428191  
14                      The Beaches West,India Bazaar  43.668999 -79.315572  
15                      Commerce Court,Victoria Hotel  43.648198 -79.379817  
16                                    Studio District  43.659526 -79.340923  
17                                      Lawrence Park  43.728020 -79.388790  
18                                           Roselawn  43.711695 -79.416936  
19                                   Davisville North  43.712751 -79.390197  
20                 Forest Hill North,Forest Hill West  43.696948 -79.411307  
21                       High Park,The Junction South  43.661608 -79.464763  
22                                 North Toronto West  43.715383 -79.405678  
23                  The Annex,North Midtown,Yorkville  43.672710 -79.405678  
24                              Parkdale,Roncesvalles  43.648960 -79.456325  
25                                         Davisville  43.704324 -79.388790  
26                      Harbord,University of Toronto  43.662696 -79.400049  
27                                  Runnymede,Swansea  43.651571 -79.484450  
28                         Moore Park,Summerhill East  43.689574 -79.383160  
29            Chinatown,Grange Park,Kensington Market  43.653206 -79.400049  
30  Deer Park,Forest Hill SE,Rathnelly,South Hill,...  43.686412 -79.400049  
31  CN Tower,Bathurst Quay,Island airport,Harbourf...  43.628947 -79.394420  
32                                           Rosedale  43.67956

### 3.2.2. Visualization of the neighborhoods to segment and cluster

As we did with all of Toronto City, let's visualizate the neighborhoods of these four boroughs.

In [35]:
address = 'Toronto'

geolocator = Nominatim(user_agent = "toronto_explorer")
location   = geolocator.geocode(address)
latitude   = location.latitude
longitude  = location.longitude

print('The geographical coordinates of the city of Toronto are latitude = {} and longitude = {}.'.format(latitude, longitude))

The geographical coordinates of the city of Toronto are latitude = 43.653963 and longitude = -79.387207.


In [36]:
# create map of Toronto using latitude and longitude values
map_toronto_cp = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_t_cp['Latitude'],df_t_cp['Longitude'], df_t_cp['Borough'], df_t_cp['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto_cp)  
    
map_toronto_cp

### 3.2.3. Explore the neighborhoods and segment them

**3.2.3.1. Define Foursquare Credentials and Version**

In [37]:
CLIENT_ID     = 'F1YOZLYDP201ASDXJCR0CSSV5PTIF2R0JHMXUU0RNA5HISCN' # your Foursquare ID
CLIENT_SECRET = 'JKJ1UD25OVAH2SR1MMNRTZPST5XCWVTRL3D4JDC3PFC3JKFS' # your Foursquare Secret
VERSION       = '20190416' # Foursquare API version

**3.2.3.2. Let's get the top 100 venues that are in each neighborhood in the dataframe within a radius of 500 meters.**

Let's create a function for this process.

In [38]:
def getNearbyVenues(bor,names, latitudes, longitudes, radius=500, LIMIT=100):
    
    venues_list=[]
    for bor,name, lat, lng in zip(bor, names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            bor,
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Borough','Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Run the above function on each neighborhood and create a new dataframe called *toronto_cp_venues*.

In [39]:
toronto_cp_venues = getNearbyVenues(bor       = df_t_cp['Borough'], 
                                    names     = df_t_cp['Neighborhood'],
                                   latitudes  = df_t_cp['Latitude'],
                                   longitudes = df_t_cp['Longitude'])

Harbourfront,Regent Park
Ryerson,Garden District
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Adelaide,King,Richmond
Dovercourt Village,Dufferin
Harbourfront East,Toronto Islands,Union Station
Little Portugal,Trinity
The Danforth West,Riverdale
Design Exchange,Toronto Dominion Centre
Brockton,Exhibition Place,Parkdale Village
The Beaches West,India Bazaar
Commerce Court,Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North,Forest Hill West
High Park,The Junction South
North Toronto West
The Annex,North Midtown,Yorkville
Parkdale,Roncesvalles
Davisville
Harbord,University of Toronto
Runnymede,Swansea
Moore Park,Summerhill East
Chinatown,Grange Park,Kensington Market
Deer Park,Forest Hill SE,Rathnelly,South Hill,Summerhill West
CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara
Rosedale
Stn A PO Boxes 25 The Esplanade
Cabbagetown,St. James Town
First Canadian Place,Underground city


Let's check the resulting dataframe:

In [40]:
print(toronto_cp_venues.shape)
toronto_cp_venues.head()

(1702, 8)


Borough              Neighborhood  Neighborhood Latitude  \
0  Downtown Toronto  Harbourfront,Regent Park               43.65426   
1  Downtown Toronto  Harbourfront,Regent Park               43.65426   
2  Downtown Toronto  Harbourfront,Regent Park               43.65426   
3  Downtown Toronto  Harbourfront,Regent Park               43.65426   
4  Downtown Toronto  Harbourfront,Regent Park               43.65426   

   Neighborhood Longitude                                            Venue  \
0              -79.360636                                 Roselle Desserts   
1              -79.360636                                    Tandem Coffee   
2              -79.360636  Toronto Cooper Koo Family Cherry St YMCA Centre   
3              -79.360636                               Morning Glory Cafe   
4              -79.360636                              Body Blitz Spa East   

   Venue Latitude  Venue Longitude        Venue Category  
0       43.653447       -79.362017                Bakery  
1       43.653559       -79.361809           Coffee Shop  
2       43.653191       -79.357947  Gym / Fitness Center  
3       43.653947       -79.361149        Breakfast Spot  
4       43.654735       -79.359874                   Spa

Let's check how many venues were returned for each borough:

In [41]:
toronto_cp_venues.groupby('Borough').count()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
Borough                                                                         
Central Toronto            112                    112                     112   
Downtown Toronto          1294                   1294                    1294   
East Toronto               117                    117                     117   
West Toronto               179                    179                     179   

                  Venue  Venue Latitude  Venue Longitude  Venue Category  
Borough                                                                   
Central Toronto     112             112              112             112  
Downtown Toronto   1294            1294             1294            1294  
East Toronto        117             117              117             117  
West Toronto        179             179              179             179

Let's check how many venues categories are:

In [42]:
print('There are {} uniques categories.'.format(len(toronto_cp_venues['Venue Category'].unique())))

There are 233 uniques categories.


Let's check the distribution of these venues categories: 

In [43]:
toronto_cp_venues.groupby(['Borough', 'Neighborhood'])[['Venue Category']].count()

Venue Category
Borough          Neighborhood                                                      
Central Toronto  Davisville                                                      34
                 Davisville North                                                 9
                 Deer Park,Forest Hill SE,Rathnelly,South Hill,S...              14
                 Forest Hill North,Forest Hill West                               4
                 Lawrence Park                                                    3
                 Moore Park,Summerhill East                                       3
                 North Toronto West                                              20
                 Roselawn                                                         1
                 The Annex,North Midtown,Yorkville                               24
Downtown Toronto Adelaide,King,Richmond                                         100
                 Berczy Park                                                     56
                 CN Tower,Bathurst Quay,Island airport,Harbourfr...              15
                 Cabbagetown,St. James Town                                      48
                 Central Bay Street                                              86
                 Chinatown,Grange Park,Kensington Market                        100
                 Christie                                                        16
                 Church and Wellesley                                            89
                 Commerce Court,Victoria Hotel                                  100
                 Design Exchange,Toronto Dominion Centre                        100
                 First Canadian Place,Underground city                          100
                 Harbord,University of Toronto                                   34
                 Harbourfront East,Toronto Islands,Union Station                100
                 Harbourfront,Regent Park                                        50
                 Rosedale                                                         4
                 Ryerson,Garden District                                        100
                 St. James Town                                                 100
                 Stn A PO Boxes 25 The Esplanade                                 96
East Toronto     Business Reply Mail Processing Centre 969 Eastern               17
                 Studio District                                                 37
                 The Beaches                                                      4
                 The Beaches West,India Bazaar                                   16
                 The Danforth West,Riverdale                                     43
West Toronto     Brockton,Exhibition Place,Parkdale Village                      22
                 Dovercourt Village,Dufferin                                     18
                 High Park,The Junction South                                    24
                 Little Portugal,Trinity                                         62
                 Parkdale,Roncesvalles                                           16
                 Runnymede,Swansea                                               37

**3.2.3.3. Analyze each neighborhood.**

In [44]:
# one hot encoding
toronto_cp_onehot = pd.get_dummies(toronto_cp_venues[['Venue Category']], prefix="", prefix_sep="")

# add borough and neighborhood column back to dataframe
toronto_cp_onehot['Borough'] = toronto_cp_venues['Borough'] 
toronto_cp_onehot['Neighborhood'] = toronto_cp_venues['Neighborhood'] 

# move borough and neighborhood columns to the first two columns
borough = toronto_cp_onehot['Borough']
toronto_cp_onehot.drop(labels=['Borough'], axis=1,inplace = True)
toronto_cp_onehot.insert(0,'Borough', borough)

neigh = toronto_cp_onehot['Neighborhood']
toronto_cp_onehot.drop(labels=['Neighborhood'], axis=1,inplace = True)
toronto_cp_onehot.insert(1,'Neighborhood', neigh)

toronto_cp_onehot.head()

Borough              Neighborhood  Adult Boutique  \
0  Downtown Toronto  Harbourfront,Regent Park               0   
1  Downtown Toronto  Harbourfront,Regent Park               0   
2  Downtown Toronto  Harbourfront,Regent Park               0   
3  Downtown Toronto  Harbourfront,Regent Park               0   
4  Downtown Toronto  Harbourfront,Regent Park               0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  ...  \
0               0                0                 0  ...   
1               0                0                 0  ...   
2               0                0                 0  ...   
3               0                0                 0  ...   
4               0                0                 0  ...   

   Thrift / Vintage Store  Toy / Game Store  Trail  Train Station  \
0                       0                 0      0              0   
1                       0                 0      0              0   
2                       0                 0      0              0   
3                       0                 0      0              0   
4                       0                 0      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Wine Bar  Wings Joint  Yoga Studio  
0         0            0            0  
1         0            0            0  
2         0            0            0  
3         0            0            0  
4         0            0            0  

[5 rows x 234 columns]

Let's examine the new dataframe size:

In [45]:
toronto_cp_onehot.shape

(1702, 234)

Let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category:

In [46]:
toronto_cp_grouped = toronto_cp_onehot.groupby(['Borough','Neighborhood']).mean().reset_index()
toronto_cp_grouped

Borough                                       Neighborhood  \
0    Central Toronto                                         Davisville   
1    Central Toronto                                   Davisville North   
2    Central Toronto  Deer Park,Forest Hill SE,Rathnelly,South Hill,...   
3    Central Toronto                 Forest Hill North,Forest Hill West   
4    Central Toronto                                      Lawrence Park   
5    Central Toronto                         Moore Park,Summerhill East   
6    Central Toronto                                 North Toronto West   
7    Central Toronto                                           Roselawn   
8    Central Toronto                  The Annex,North Midtown,Yorkville   
9   Downtown Toronto                             Adelaide,King,Richmond   
10  Downtown Toronto                                        Berczy Park   
11  Downtown Toronto  CN Tower,Bathurst Quay,Island airport,Harbourf...   
12  Downtown Toronto                         Cabbagetown,St. James Town   
13  Downtown Toronto                                 Central Bay Street   
14  Downtown Toronto            Chinatown,Grange Park,Kensington Market   
15  Downtown Toronto                                           Christie   
16  Downtown Toronto                               Church and Wellesley   
17  Downtown Toronto                      Commerce Court,Victoria Hotel   
18  Downtown Toronto            Design Exchange,Toronto Dominion Centre   
19  Downtown Toronto              First Canadian Place,Underground city   
20  Downtown Toronto                      Harbord,University of Toronto   
21  Downtown Toronto    Harbourfront East,Toronto Islands,Union Station   
22  Downtown Toronto                           Harbourfront,Regent Park   
23  Downtown Toronto                                           Rosedale   
24  Downtown Toronto                            Ryerson,Garden District   
25  Downtown Toronto                                     St. James Town   
26  Downtown Toronto                    Stn A PO Boxes 25 The Esplanade   
27      East Toronto  Business Reply Mail Processing Centre 969 Eastern   
28      East Toronto                                    Studio District   
29      East Toronto                                        The Beaches   
30      East Toronto                      The Beaches West,India Bazaar   
31      East Toronto                        The Danforth West,Riverdale   
32      West Toronto         Brockton,Exhibition Place,Parkdale Village   
33      West Toronto                        Dovercourt Village,Dufferin   
34      West Toronto                       High Park,The Junction South   
35      West Toronto                            Little Portugal,Trinity   
36      West Toronto                              Parkdale,Roncesvalles   
37      West Toronto                                  Runnymede,Swansea   

    Adult Boutique  Afghan Restaurant   Airport  Airport Food Court  \
0         0.000000           0.000000  0.000000            0.000000   
1         0.000000           0.000000  0.000000            0.000000   
2         0.000000           0.000000  0.000000            0.000000   
3         0.000000           0.000000  0.000000            0.000000   
4         0.000000           0.000000  0.000000            0.000000   
5         0.000000           0.000000  0.000000            0.000000   
6         0.000000           0.000000  0.000000            0.000000   
7         0.000000           0.000000  0.000000            0.000000   
8         0.000000           0.000000  0.000000            0.000000   
9         0.000000           0.000000  0.000000            0.000000   
10        0.000000           0.000000  0.000000            0.000000   
11        0.000000           0.000000  0.066667            0.066667   
12        0.000000           0.000000  0.000000            0.000000   
13        0.000000           0.000000  0.000000            0.000000   
14        0.000000    

Let's confirm the new size:

In [47]:
toronto_cp_grouped.shape

(38, 234)

Let's print each neighborhood along with the top 5 most common venues:

In [48]:
num_top_venues = 5

for hood in toronto_cp_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_cp_grouped[toronto_cp_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[2:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Davisville----
                venue  freq
0      Sandwich Place  0.09
1        Dessert Shop  0.09
2                Café  0.06
3         Pizza Place  0.06
4  Italian Restaurant  0.06


----Davisville North----
            venue  freq
0  Breakfast Spot  0.11
1           Hotel  0.11
2            Park  0.11
3             Gym  0.11
4      Restaurant  0.11


----Deer Park,Forest Hill SE,Rathnelly,South Hill,Summerhill West----
                 venue  freq
0                  Pub  0.14
1          Coffee Shop  0.14
2          Pizza Place  0.07
3  American Restaurant  0.07
4     Sushi Restaurant  0.07


----Forest Hill North,Forest Hill West----
                venue  freq
0  Mexican Restaurant  0.25
1               Trail  0.25
2    Sushi Restaurant  0.25
3       Jewelry Store  0.25
4      Adult Boutique  0.00


----Lawrence Park----
            venue  freq
0            Park  0.33
1     Swim School  0.33
2        Bus Line  0.33
3  Adult Boutique  0.00
4          Museum  0.00


----Moore Par

Let's put that into a *pandas* dataframe.

1) First, let's write a function to sort the venues in descending order:

In [49]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[2:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

2) Let's create the new dataframe and display the top 10 venues for each neighborhood:

In [50]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Borough', 'Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Borough'] = toronto_cp_grouped['Borough']
neighborhoods_venues_sorted['Neighborhood'] = toronto_cp_grouped['Neighborhood']


for ind in np.arange(toronto_cp_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 2:] = return_most_common_venues(toronto_cp_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Borough                                       Neighborhood  \
0    Central Toronto                                         Davisville   
1    Central Toronto                                   Davisville North   
2    Central Toronto  Deer Park,Forest Hill SE,Rathnelly,South Hill,...   
3    Central Toronto                 Forest Hill North,Forest Hill West   
4    Central Toronto                                      Lawrence Park   
5    Central Toronto                         Moore Park,Summerhill East   
6    Central Toronto                                 North Toronto West   
7    Central Toronto                                           Roselawn   
8    Central Toronto                  The Annex,North Midtown,Yorkville   
9   Downtown Toronto                             Adelaide,King,Richmond   
10  Downtown Toronto                                        Berczy Park   
11  Downtown Toronto  CN Tower,Bathurst Quay,Island airport,Harbourf...   
12  Downtown Toronto                         Cabbagetown,St. James Town   
13  Downtown Toronto                                 Central Bay Street   
14  Downtown Toronto            Chinatown,Grange Park,Kensington Market   
15  Downtown Toronto                                           Christie   
16  Downtown Toronto                               Church and Wellesley   
17  Downtown Toronto                      Commerce Court,Victoria Hotel   
18  Downtown Toronto            Design Exchange,Toronto Dominion Centre   
19  Downtown Toronto              First Canadian Place,Underground city   
20  Downtown Toronto                      Harbord,University of Toronto   
21  Downtown Toronto    Harbourfront East,Toronto Islands,Union Station   
22  Downtown Toronto                           Harbourfront,Regent Park   
23  Downtown Toronto                                           Rosedale   
24  Downtown Toronto                            Ryerson,Garden District   
25  Downtown Toronto                                     St. James Town   
26  Downtown Toronto                    Stn A PO Boxes 25 The Esplanade   
27      East Toronto  Business Reply Mail Processing Centre 969 Eastern   
28      East Toronto                                    Studio District   
29      East Toronto                                        The Beaches   
30      East Toronto                      The Beaches West,India Bazaar   
31      East Toronto                        The Danforth West,Riverdale   
32      West Toronto         Brockton,Exhibition Place,Parkdale Village   
33      West Toronto                        Dovercourt Village,Dufferin   
34      West Toronto                       High Park,The Junction South   
35      West Toronto                            Little Portugal,Trinity   
36      West Toronto                              Parkdale,Roncesvalles   
37      West Toronto                                  Runnymede,Swansea   

   1st Most Common Venue          2nd Most Common Venue 3rd Most Common Venue  \
0         Sandwich Place                   Dessert Shop      Sushi Restaurant   
1                    Gym              Food & Drink Shop                  Park   
2                    Pub                    Coffee Shop     Convenience Store   
3                  Trail             Mexican Restaurant         Jewelry Store   
4               Bus Line                           Park           Swim School   
5             Restaurant                            Gym            Playground   
6         Clothing Store                    Coffee Shop   Sporting Goods Shop   
7                 Garden                    Yoga Studio    Falafel Restaurant   
8            Coffee Shop                 Sandwich Place                  Café   
9            Coffee Shop                Thai Restaurant                  Café   
10           Coffee Shop                   Cocktail Bar    Seafood Restaurant   
11       Airport Service               Airport Terminal        Airport Lounge   
12           Coffee Shop          

**3.2.3.4. Cluster neighborhoods.**

Run *k*-means to cluster the neighborhood into 5 clusters:

In [51]:
# set number of clusters
kclusters = 5

toronto_cp_grouped_clustering = toronto_cp_grouped.drop(['Borough','Neighborhood'], 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_cp_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 4, 3, 0, 1, 0, 0], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood:

In [52]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_cp_merged = df_t_cp 

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_cp_merged = toronto_cp_merged.merge(neighborhoods_venues_sorted, on='Neighborhood')
toronto_cp_merged = toronto_cp_merged.drop(['PostalCode','Borough_y'], 1)
toronto_cp_merged = toronto_cp_merged.rename(columns = {"Borough_x": "Borough"}) 
toronto_cp_merged 

#toronto_cp_merged.head() # check the last columns!

Borough                                       Neighborhood  \
0   Downtown Toronto                           Harbourfront,Regent Park   
1   Downtown Toronto                            Ryerson,Garden District   
2   Downtown Toronto                                     St. James Town   
3       East Toronto                                        The Beaches   
4   Downtown Toronto                                        Berczy Park   
5   Downtown Toronto                                 Central Bay Street   
6   Downtown Toronto                                           Christie   
7   Downtown Toronto                             Adelaide,King,Richmond   
8       West Toronto                        Dovercourt Village,Dufferin   
9   Downtown Toronto    Harbourfront East,Toronto Islands,Union Station   
10      West Toronto                            Little Portugal,Trinity   
11      East Toronto                        The Danforth West,Riverdale   
12  Downtown Toronto            Design Exchange,Toronto Dominion Centre   
13      West Toronto         Brockton,Exhibition Place,Parkdale Village   
14      East Toronto                      The Beaches West,India Bazaar   
15  Downtown Toronto                      Commerce Court,Victoria Hotel   
16      East Toronto                                    Studio District   
17   Central Toronto                                      Lawrence Park   
18   Central Toronto                                           Roselawn   
19   Central Toronto                                   Davisville North   
20   Central Toronto                 Forest Hill North,Forest Hill West   
21      West Toronto                       High Park,The Junction South   
22   Central Toronto                                 North Toronto West   
23   Central Toronto                  The Annex,North Midtown,Yorkville   
24      West Toronto                              Parkdale,Roncesvalles   
25   Central Toronto                                         Davisville   
26  Downtown Toronto                      Harbord,University of Toronto   
27      West Toronto                                  Runnymede,Swansea   
28   Central Toronto                         Moore Park,Summerhill East   
29  Downtown Toronto            Chinatown,Grange Park,Kensington Market   
30   Central Toronto  Deer Park,Forest Hill SE,Rathnelly,South Hill,...   
31  Downtown Toronto  CN Tower,Bathurst Quay,Island airport,Harbourf...   
32  Downtown Toronto                                           Rosedale   
33  Downtown Toronto                    Stn A PO Boxes 25 The Esplanade   
34  Downtown Toronto                         Cabbagetown,St. James Town   
35  Downtown Toronto              First Canadian Place,Underground city   
36  Downtown Toronto                               Church and Wellesley   
37      East Toronto  Business Reply Mail Processing Centre 969 Eastern   

     Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0   43.654260 -79.360636               0           Coffee Shop   
1   43.657162 -79.378937               0           Coffee Shop   
2   43.651494 -79.375418               0           Coffee Shop   
3   43.676357 -79.293031               0     Health Food Store   
4   43.644771 -79.373306               0           Coffee Shop   
5   43.657952 -79.387383               0           Coffee Shop   
6   43.669542 -79.422564               0         Grocery Store   
7   43.650571 -79.384568               0           Coffee Shop   
8   43.669005 -79.442259               0           Supermarket   
9   43.640816 -79.381752               0           Coffee Shop   
10  43.647927 -79.419750               0                   Bar   
11  43.679557 -79.352188               0      Greek Restaurant   
12  43.647177 -79.381576               0           Coffee Shop   
13  43.636847 -79.428191               0        Breakfast Spot   
14  43.668999 -79.315572               0             Pet Store   
15  43.648198 -79.379817       

Finally, let's visualize the resulting clusters:

In [53]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, bor, cluster in zip(toronto_cp_merged['Latitude'], toronto_cp_merged['Longitude'], toronto_cp_merged['Neighborhood'], toronto_cp_merged['Borough'], toronto_cp_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ', ' + bor + ', ' + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

**3.2.3.5. Examine clusters.**

**Cluster 0**

In [54]:
toronto_cp_merged.loc[toronto_cp_merged['Cluster Labels'] == 0, toronto_cp_merged.columns[[1] + list(range(5,toronto_cp_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
0                            Harbourfront,Regent Park           Coffee Shop   
1                             Ryerson,Garden District           Coffee Shop   
2                                      St. James Town           Coffee Shop   
3                                         The Beaches     Health Food Store   
4                                         Berczy Park           Coffee Shop   
5                                  Central Bay Street           Coffee Shop   
6                                            Christie         Grocery Store   
7                              Adelaide,King,Richmond           Coffee Shop   
8                         Dovercourt Village,Dufferin           Supermarket   
9     Harbourfront East,Toronto Islands,Union Station           Coffee Shop   
10                            Little Portugal,Trinity                   Bar   
11                        The Danforth West,Riverdale      Greek Restaurant   
12            Design Exchange,Toronto Dominion Centre           Coffee Shop   
13         Brockton,Exhibition Place,Parkdale Village        Breakfast Spot   
14                      The Beaches West,India Bazaar             Pet Store   
15                      Commerce Court,Victoria Hotel           Coffee Shop   
16                                    Studio District                  Café   
19                                   Davisville North                   Gym   
20                 Forest Hill North,Forest Hill West                 Trail   
21                       High Park,The Junction South                   Bar   
22                                 North Toronto West        Clothing Store   
23                  The Annex,North Midtown,Yorkville           Coffee Shop   
24                              Parkdale,Roncesvalles        Breakfast Spot   
25                                         Davisville        Sandwich Place   
26                      Harbord,University of Toronto                  Café   
27                                  Runnymede,Swansea           Coffee Shop   
29            Chinatown,Grange Park,Kensington Market                  Café   
30  Deer Park,Forest Hill SE,Rathnelly,South Hill,...                   Pub   
31  CN Tower,Bathurst Quay,Island airport,Harbourf...       Airport Service   
33                    Stn A PO Boxes 25 The Esplanade           Coffee Shop   
34                         Cabbagetown,St. James Town           Coffee Shop   
35              First Canadian Place,Underground city           Coffee Shop   
36                               Church and Wellesley           Coffee Shop   
37  Business Reply Mail Processing Centre 969 Eastern    Light Rail Station   

            2nd Most Common Venue 3rd Most Common Venue  \
0                          Bakery                  Café   
1                  Clothing Store                  Café   
2                      Restaurant                  Café   
3                     Coffee Shop                   Pub   
4                    Cocktail Bar    Seafood Restaurant   
5              Italian Restaurant       Bubble Tea Shop   
6                            Café                  Park   
7                 Thai Restaurant                  Café   
8                          Bakery              Pharmacy   
9                        Aquarium                 Hotel   
10                    Coffee Shop      Asian Restaurant   
11                    Coffee Shop        Ice Cream Shop   
12                           Café                 Hotel   
13                           Café           Coffee Shop   
14             Italian Restaurant            Steakhouse   
15                          Hotel                  Café   
16                    Coffee Shop                Bakery   
19              Food & Drink Shop                  Park   
20             Mexican Restaurant         Jewelry Store   
21             Mexican Restaurant                  Café   
22                    Coffee Shop   

**Cluster 1**

In [55]:
toronto_cp_merged.loc[toronto_cp_merged['Cluster Labels'] == 1, toronto_cp_merged.columns[[1] + list(range(5,toronto_cp_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
18     Roselawn                Garden           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
18    Falafel Restaurant           Event Space  Ethiopian Restaurant   

   6th Most Common Venue        7th Most Common Venue 8th Most Common Venue  \
18     Electronics Store  Eastern European Restaurant   Dumpling Restaurant   

   9th Most Common Venue 10th Most Common Venue  
18            Donut Shop       Doner Restaurant

**Cluster 2**

In [56]:
toronto_cp_merged.loc[toronto_cp_merged['Cluster Labels'] == 2, toronto_cp_merged.columns[[1] + list(range(5,toronto_cp_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
32     Rosedale                  Park            Playground   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
32                 Trail           Yoga Studio          Dessert Shop   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
32           Event Space  Ethiopian Restaurant     Electronics Store   

          9th Most Common Venue 10th Most Common Venue  
32  Eastern European Restaurant    Dumpling Restaurant

**Cluster 3**

In [57]:
toronto_cp_merged.loc[toronto_cp_merged['Cluster Labels'] == 3, toronto_cp_merged.columns[[1] + list(range(5,toronto_cp_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
28  Moore Park,Summerhill East            Restaurant                   Gym   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
28            Playground      Department Store           Event Space   

   6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
28  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   

   9th Most Common Venue 10th Most Common Venue  
28   Dumpling Restaurant             Donut Shop

**Cluster 4**

In [58]:
toronto_cp_merged.loc[toronto_cp_merged['Cluster Labels'] == 4, toronto_cp_merged.columns[[1] + list(range(5,toronto_cp_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
17  Lawrence Park              Bus Line                  Park   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
17           Swim School           Yoga Studio    Falafel Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
17           Event Space  Ethiopian Restaurant     Electronics Store   

          9th Most Common Venue 10th Most Common Venue  
17  Eastern European Restaurant    Dumpling Restaurant